src/registration.rs

```diff
@@ -100,46 +92,126 @@ impl Registrar {
+                );
+
+            rx.recv().unwrap().unwrap();
+            self.certificate_manager.reload().unwrap();
+        }
+    }
+
+
+    pub fn start<T: Controller>(self,
+                                iface: Option<String>,
+                                tunnel: &Option<Tunnel>,
+                                box_port: u16,
+                                controller: &T) {
+        info!("Starting registration with {}", self.registration_endpoint);
+
+        let ip_addr = self.get_ip_addr(&iface);
+        if ip_addr == None {
+            // TODO: retry later, in case we're racing with the network
+            // configuration.
+            return;
+        }
+
+        info!("Got ip address: {}", ip_addr.clone().unwrap());
+
+        let tunnel_frontend = if let Some(ref tunnel) = *tunnel {
+            tunnel.get_frontend_name()
+        } else {
+            None
+        };
+
+        let http_scheme = if controller.get_tls_enabled() {
+            "https"
+        } else {
+            "http"
+        };
+
+
+        let enabled_tls = controller.get_tls_enabled();
 
         // Spawn a thread to register every REGISTRATION_INTERVAL_IN_MINUTES.
         thread::Builder::new().name("Registrar".to_owned())
-                              .spawn(move || {
-            loop {
-                let client = Client::new();
-                let res = client.post(&endpoint_url)
-                    .header(Connection::close())
-                    .body(&body)
-                    .send();
-
-                // Sanity checks, mostly to debug errors since we don't try
-                // to recover from failures.
-                if let Ok(mut response) = res {
-                    if response.status == StatusCode::Ok {
-                        let mut body = String::new();
-                        if let Ok(_) = response.read_to_string(&mut body) {
-                            info!("Server responded with: {}", body);
-                        } else {
-                            info!("Unable to read answer from {}", endpoint_url);
-                        }
-                    }
-                } else {
-                    info!("Unable to send request to {}", endpoint_url);
+            .spawn(move || {
+                let tunnel_configured = tunnel_frontend.clone().is_some();
+
+                if enabled_tls {
+                    self.register_certificates();
                 }
 
-                // Create entry for local DNS
-                certificate_manager.register_dns_record(
-                    "A",
-                    &format!("local.{}", domain),
-                    &ip_addr.clone().unwrap(),
-                    &dns_api_endpoint.clone(),
-                    &box_hostname
-                ).unwrap();
-
-                // Go to sleep.
-                thread::sleep(Duration::from_secs(REGISTRATION_INTERVAL_IN_MINUTES as u64 * 60))
-            }
-        }).unwrap();
+                loop {
```

Comment:
We also need to update the registration info (specially the DNS records) if the local IP change. File a follow up for that, please.